alma/client.py

```diff
@@ -1,25 +1,80 @@
+from .credentials import MerchantIdCredentials, AlmaSessionCredentials, ApiKeyCredentials
 from .version import __version__ as alma_version
 
 
 class Client:
     SANDBOX_API_URL = "https://api.sandbox.getalma.eu"
     LIVE_API_URL = "https://api.getalma.eu"
 
-    def __init__(self, api_key, **options):
-        if not api_key:
-            raise ValueError("An API key is required to instantiate a new Client")
+    @classmethod
+    def with_api_key(cls, api_key: str, **options):
+        options = {**options, "credentials": ApiKeyCredentials(api_key)}
+        return cls(None, **options)
+
+    @classmethod
+    def with_merchant_id(cls, merchant_id: str, **options):
+        options = {**options, "credentials": MerchantIdCredentials(merchant_id)}
+        return cls(None, **options)
+
+    @classmethod
+    def with_alma_session(cls, session_id: str, cookie_name: str = "alma_sess", **options):
+        options = {**options, "credentials": AlmaSessionCredentials(session_id, cookie_name)}
+        return cls(None, **options)
+
+    def __init__(self, api_key: Optional[str] = None, **kwargs):
+        """
+        Create a new instance of the Alma API Client.
+
+        It is recommended to use one the convenience methods instead of the default constructor:
+        - Client.with_api_key
+        - Client.with_merchant_id
+        - Client.with_alma_session
+
+        :param  api_key:    Deprecated - use Client.with_api_key("<api_key>") instead
+        :type   api_key:    str
+
+        :keyword    credentials A `Credentials` instance to be used to configure requests made to the API
+                                This would typically be set by one of the convenience methods mentioned above
+        :keyword    mode        API mode to be used: either ApiModes.LIVE or ApiModes.TEST
+        :keyword    logger      A logger instance to be used instead of the default one
+        :keyword    api_root    root URL(s) to call Alma's API at. You probably don't want to change it!
+
+                                Expected types:
+                                -------------
+                                str: the provided URL will be used for both LIVE and TEST modes
+                                dict: must have two keys, ApiModes.LIVE and ApiModes.TEST, each value must be
+                                      the URL to be used for each mode
+
+        """
+        default_mode = ApiModes.LIVE
+        credentials = kwargs.get("credentials")
+
+        if isinstance(credentials, ApiKeyCredentials):
+            api_key = credentials.api_key
+
+        if api_key:
+            default_mode = ApiModes.LIVE if api_key.startswith("sk_live") else ApiModes.TEST
+
+            # Backward compatibility with the older init method
+            if not credentials:
+                credentials = ApiKeyCredentials(api_key)
+
+        if not credentials:
+            raise ValueError("Valid credentials are required to instantiate a new Client")
```

Comment:
Avec une property `mode` sur `Credentials` qui renvoie `ApiModes.LIVE`, redéfinie dans `ApiKeyCredentials`, tu pourrais simplifier ce constructeur:
```suggestion
        if not credentials:
            if not api_key:
                raise ValueError("Valid credentials are required to instantiate a new Client")
            # Backward compatibility with the older init method
            credentials = ApiKeyCredentials(api_key)

        default_mode = credentials.mode
```